### PR TITLE
Fix #45305 + add regression test GAS

### LINKED ISSUE #45305: Gradients not averaged by GAS when using DeepSpeed + model_accepts_loss_kwargs=True (Qwen3, Llama3, etc.)
### System Info

- `transformers` version: 5.3.0
- Platform: Linux-5.14.0-427.33.1.el9_4.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.15
- Huggingface_hub version: 1.6.0
- Safetensors version: 0.7.0
- Accelerate version: 1.13.0
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: True
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- DeepSpeed version: 0.18.9
- PyTorch version (accelerator?): 2.6.0+cu124 (CUDA)
- Using distributed or parallel set-up in script?: Yes
- Using GPU in script?: Yes
- GPU type: NVIDIA L40S

### Who can help?

@SunMarc, @3outeille, @ArthurZucker 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

## Bug description
 
When using the HuggingFace `Trainer` with DeepSpeed and any modern model where `model_accepts_loss_kwargs=True` (e.g., Qwen3, Llama3, and most recent HuggingFace models), gradients are **summed instead of averaged** across gradient accumulation steps. This causes ~GAS× higher grad norms compared to training the same model **without** DeepSpeed, leading to divergent training dynamics.
 
## Root cause
 
The issue is a conflict between two code paths in `Trainer.training_step()` (`trainer.py`, lines 1925–1933 in v5.3.0):
 
```python
# Line 1925-1927: Conditional loss normalization
if (not self.model_accepts_loss_kwargs or num_items_in_batch is None) and self.compute_loss_func is None:
    loss = loss / self.current_gradient_accumulation_steps
 
# Line 1931-1933: Unconditional scale_wrt_gas=False for DeepSpeed (added by PR #35808)
if self.accelerator.distributed_type == DistributedType.DEEPSPEED:
    kwargs["scale_wrt_gas"] = False
 
self.accelerator.backward(loss, **kwargs)
```
 
**PR [#35808](https://github.com/huggingface/transformers/pull/35808)** added `scale_wrt_gas=False` to prevent double-scaling when the Trainer already divides loss by GAS. This fix is correct when `model_accepts_loss_kwargs=False` (Trainer divides on line 1927, and `scale_wrt_gas=False` prevents DeepSpeed from dividing again).
 
However, for modern models where `model_accepts_loss_kwargs=True`:
1. **Trainer skips dividing** (line 1925 condition is False)
2. **DeepSpeed's gradient scaling is disabled** (`scale_wrt_gas=False` → `_backward_prologue_per_tensor` hook is a no-op)
3. **Nobody divides by GAS** → gradients are summed, not averaged
 
### Trace of the bug
 
| Step | `model_accepts_loss_kwargs=True` (Qwen3, etc.) | `model_accepts_loss_kwargs=False` (older models) |
|---|---|---|
| Trainer divides loss by GAS? | ❌ No (line 1925 skipped) | ✅ Yes (line 1927) |
| `scale_wrt_gas=False` passed? | ✅ Yes (line 1932) | ✅ Yes (line 1932) |
| DeepSpeed hook divides grads? | ❌ No (`_scale_wrt_gas=False`) | ❌ No (`_scale_wrt_gas=False`) |
| **Net GAS scaling** | **❌ None — gradients summed** | **✅ 1/GAS from Trainer — correct** |
 
## Minimal reproduction
 
**Important:** Run each configuration as a **separate process** so the model is freshly loaded each time.
 
```python
"""
min_repro.py — Run separately for each configuration:
  python min_repro.py --use_ds false
  deepspeed --num_gpus=1 min_repro.py --use_ds true
 
Compare grad_norm at equivalent steps between the two runs.
With GAS=8 and the bug present, DS grad_norm will be ~8x higher.
"""
import argparse, json, os, random, tempfile
import torch
from datasets import Dataset
from transformers import (
    AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments, set_seed,
)
 
DS_CONFIG = {
    "bf16": {"enabled": "auto"},
    "zero_optimization": {"stage": 3,
        "offload_optimizer": {"device": "cpu", "pin_memory": True},
        "offload_param": {"device": "cpu", "pin_memory": True},
        "overlap_comm": True, "contiguous_gradients": True,
        "reduce_bucket_size": "auto",
        "stage3_prefetch_bucket_size": "auto",
        "stage3_param_persistence_threshold": "auto"},
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
}
 
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--use_ds", type=str, default="false")
    parser.add_argument("--local_rank", type=int, default=-1)
    args = parser.parse_args()
    use_ds = args.use_ds.lower() in ("true", "1", "yes")
 
    set_seed(42)
 
    model_name = "Qwen/Qwen3-0.6B-Base"
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
 
    model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.bfloat16)
    model.resize_token_embeddings(len(tokenizer))
 
    # Random tokens so loss does not trivially collapse
    random.seed(42)
    vocab, seq_len, n = len(tokenizer), 128, 200
    data = {
        "input_ids": [[random.randint(0, vocab-1) for _ in range(seq_len)] for _ in range(n)],
        "labels":    [[random.randint(0, vocab-1) for _ in range(seq_len)] for _ in range(n)],
    }
    dataset = Dataset.from_dict(data)
 
    ds_path = None
    if use_ds:
        ds_path = os.path.join(tempfile.mkdtemp(), "ds.json")
        with open(ds_path, "w") as f: json.dump(DS_CONFIG, f)
 
    training_args = TrainingArguments(
        output_dir=f"./test_gas_{'ds' if use_ds else 'no_ds'}",
        per_device_train_batch_size=1,
        gradient_accumulation_steps=8,
        max_steps=20, logging_steps=1,
        learning_rate=5e-5, lr_scheduler_type="cosine",
        bf16=True, deepspeed=ds_path,
        report_to="none", save_strategy="no",
        max_grad_norm=0.0,  # Disable clipping so raw grad norms are visible
    )
    trainer = Trainer(model=model, args=training_args,
                      train_dataset=dataset, processing_class=tokenizer)
 
    print(f"\nDeepSpeed={use_ds}, model_accepts_loss_kwargs={trainer.model_accepts_loss_kwargs}")
    trainer.train()
 
if __name__ == "__main__":
    main()
```
 
Run as **two separate invocations** and compare the `grad_norm` column:
```bash
python min_repro.py --use_ds false          # baseline
deepspeed --num_gpus=1 min_repro.py --use_ds true   # ~8x higher grad_norm expected
```
 
## Evidence from real training
 
Training Qwen3-1.7B-Base with identical configs (GAS=8, lr=2e-5, cosine schedule, bf16), with and without DeepSpeed ZeRO-3:
 
- **train/grad_norm**: DeepSpeed run shows ~5-7× higher grad norms (compressed from theoretical 8× by `max_grad_norm=1.0` clipping)
- **train/loss**: DeepSpeed run plateaus ~0.5 higher — model underfits because effective LR is ~8× too large (unaveraged gradients × clipped norm still produce larger updates)
- **train/learning_rate**: Schedules diverge because the optimizer behaves differently under inflated gradients
 
The ratio of grad norms matches the `gradient_accumulation_steps=8` value, confirming that no GAS averaging is happening in the DeepSpeed path.
 

### Expected behavior

 
Gradient norms and loss curves should be **identical** (up to numerical precision) regardless of whether DeepSpeed is enabled, given the same effective batch size and hyperparameters.
 
The fix should condition `scale_wrt_gas` on whether the Trainer already normalized the loss:
 
```python
# Proposed fix in training_step():
if self.accelerator.distributed_type == DistributedType.DEEPSPEED:
    if not self.model_accepts_loss_kwargs or num_items_in_batch is None:
        # Trainer already divided by GAS above → disable DeepSpeed's scaling
        kwargs["scale_wrt_gas"] = False
    else:
        # Trainer did NOT divide → let DeepSpeed handle it
        kwargs["scale_wrt_gas"] = True
```
 
Alternatively, always divide in the Trainer and always set `scale_wrt_gas=False`:
 
```python
# Always normalize, regardless of model_accepts_loss_kwargs
loss = loss / self.current_gradient_accumulation_steps
 
if self.accelerator.distributed_type == DistributedType.DEEPSPEED:
    kwargs["scale_wrt_gas"] = False
```
 
## Workaround
 
Until this is fixed, users can subclass `Trainer`:
 
```python
class FixedGASTrainer(Trainer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.model_accepts_loss_kwargs = False
```
 
This forces the Trainer to always divide loss by GAS (line 1927), and combined with the existing `scale_wrt_gas=False`, produces correct gradient averaging.
 
## Related issues
 
- #35808 — PR that introduced `scale_wrt_gas=False` (correct for old models, breaks new ones)
- #40564 — Loss scales incorrectly with GAS (Gemma3 and other models)
- #34694 — Discrepancy in training loss with GA using DeepSpeed
- #37110 — Loss and gradient explosion caused by the Trainer
- DeepSpeed [#7773](https://github.com/deepspeedai/DeepSpeed/issues/7773) — Related gradient accumulation issue in pipeline parallelism
 
## Affected models
 
Any model where `model_accepts_loss_kwargs=True`, which includes most models added since late 2024: **Qwen3, Gemma3, Llama 3, Phi-4**, and any model whose `forward()` accepts `**kwargs` or has `accepts_loss_kwargs = True`.

### PR DESCRIPTION
**Pull request by florian6973**
# What does this PR do?

Fixes #45305
Add a regression test in `TrainerGradientAccumulationTest` to avoid passing the GAS value to Accelerate by mistake

Description: I force the value of the `num_steps` parameter to be 1, and the regression test just checks after the Trainer creation that the `num_steps` value is 1 and not something else.

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`[CONTRIBUTING.md](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md)`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [[contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request)](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [[forum](https://discuss.huggingface.co/)](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #45305
- [x] Did you make sure to update the documentation with your changes? *No documentation added for a bug fix, and the test is simple enough*.
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

- trainer: @SunMarc

Research projects are not maintained and should be taken as is.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -765,7 +765,8 @@ def create_accelerator_and_postprocess(self) -> None:
             else:
                 self.args.gradient_accumulation_steps = grad_acc_kwargs["num_steps"]
         else:
-            grad_acc_kwargs["num_steps"] = self.args.gradient_accumulation_steps
+            # The Trainer handles GAS itself, so GAS=1 in Accelerate to avoid any double-division
+            grad_acc_kwargs["num_steps"] = 1
```
Comment (SunMarc) [inline src/transformers/trainer.py]: I meant in the issue that this should always be the case. We don't even need the else here.

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -372,6 +372,44 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
...
+            trainer = Trainer(
+                model=model,
+                args=args,
+                train_dataset=RegressionDataset(),
+            )
+            # LlamaForCausalLM.forward() accepts **kwargs, so this must be True
+            self.assertTrue(
+                trainer.model_accepts_loss_kwargs,
+                "Test requires a model with model_accepts_loss_kwargs=True (like modern HF models).",
+            )
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: we don't need this assert

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -372,6 +372,44 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
...
 
+    def test_gradient_accumulation_steps_not_leaked_to_accelerator(self):
+        """
+        Regression test: the Trainer should not pass its gradient_accumulation_steps
+        to the Accelerator. When num_items_in_batch is active (model_accepts_loss_kwargs=True),
+        Accelerate's backward() would spuriously divide loss by GAS, resulting in
+        gradients that are GAS× too small.
+        """
+        config = LlamaConfig(vocab_size=100, hidden_size=32, num_hidden_layers=2, num_attention_heads=4)
+        model = LlamaForCausalLM(config)
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: we don't need this specific model, we can just use the usual regressionmodel i think

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -372,6 +372,44 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
...
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            args = TrainingArguments(
+                output_dir=tmp_dir,
+                per_device_train_batch_size=4,
+                gradient_accumulation_steps=4,
+                max_steps=1,
+                learning_rate=0.1,
+                report_to="none",
+                save_strategy="no",
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]:
```suggestion
```

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -372,6 +372,44 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
             # max diff broken should be very off
             self.assertGreater(max(diff_broken), 3, f"Difference {max(diff_broken)} is not greater than 3")
 
+    def test_gradient_accumulation_steps_not_leaked_to_accelerator(self):
+        """
+        Regression test: the Trainer should not pass its gradient_accumulation_steps
+        to the Accelerator. When num_items_in_batch is active (model_accepts_loss_kwargs=True),
+        Accelerate's backward() would spuriously divide loss by GAS, resulting in
+        gradients that are GAS× too small.
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: we don't need this specific description here. Maybe we can try to also add a more thorough test that actually does train the model on a few steps and compare to grad norm or the loss at the end ?

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -372,6 +372,44 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
...
+            self.assertTrue(
+                trainer.model_accepts_loss_kwargs,
+                "Test requires a model with model_accepts_loss_kwargs=True (like modern HF models).",
+            )
+            self.assertEqual(
+                trainer.accelerator.gradient_accumulation_steps,
+                1,
+                "Trainer should not leak gradient_accumulation_steps to the Accelerator. "
+                "This causes accelerator.backward() to spuriously divide loss by GAS "
+                "when num_items_in_batch is active.",
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: we don't need this extrat information here

**Review by SunMarc — COMMENTED** [review]
Thanks !! just a couple of comments

**Comment by florian6973** [comment]
Just addressed your comments, please let me know if you have any other feedback. Thank you!

**Review by SunMarc — APPROVED** [review]
Thanks a lot for iterating !
